File: src/Vector.java
/**
 * @file Vector.java
 * Represents an N-dimensional vector.
 * @author Conlan Wesson
 */

import java.util.ArrayList;

class Vector{
	private ArrayList<Integer> values = new ArrayList<Integer>(3);

	public Vector(Vector orig){
		for(int i = 0; i < orig.values.size(); i++){
			setDimension(i, orig.getDimension(i));
		}
	}

	public Vector(){
		for(int i = 0; i < values.size(); i++){
			values.set(i, 0);
		}
	}

	public Vector(int ... v) {
		for(int d = 0; d < v.length; d++){
			this.setDimension(d, v[d]);
		}
	}

	public int getDimension(int dim){
		if(dim >= values.size()){
			return 0;
		}else{
			return values.get(dim);
		}
	}

	public void setDimension(int dim, int value){
		if(dim >= values.size()){
			for(int i = values.size(); i <= dim; i++){
				values.add(i, 0);
			}
		}
		values.set(dim, value);
	}
	
	public int size(){
		return values.size();
	}

	public int getX(){
		return getDimension(0);
	}

	public int getY(){
		return getDimension(1);
	}

	public int getZ(){
		return getDimension(2);
	}

	public void setX(int value){
		setDimension(0, value);
	}

	public void setY(int value){
		setDimension(1, value);
	}

	public void setZ(int value){
		setDimension(2, value);
	}
	
	public void set(int ... dim){
		for(int i = 0; i < values.size(); i++){
			setDimension(i, 0);
		}
		for(int i = 0; i < dim.length; i++){
			setDimension(i, dim[i]);
		}
	}
	
	public void reverse(){
		for(int i = 0; i < values.size(); i++){
			values.set(i, -values.get(i));
		}
	}
	
	public void left(){
		int tx = getX();
		int ty = getY();
		setX(ty);
		setY(-tx);
	}
	
	public void right(){
		int tx = getX();
		int ty = getY();
		setX(-ty);
		setY(tx);
	}
	
	public Vector add(Vector other){
		for(int i = 0; i < this.values.size() && i < other.values.size(); i++){
			this.setDimension(i, this.getDimension(i)+other.getDimension(i));
		}
		return this;
	}
	
	@Override
	public int hashCode() {
		int hash = 1;
		for(int i : values) {
			hash *= i+1;
		}
		return hash;
	}
	
	@Override
	public boolean equals(Object obj) {
		try {
			Vector other = (Vector)obj;
			int max = Math.max(this.size(), other.size());
			for(int i = 0; i < max; i++) {
				if(this.getDimension(i) != other.getDimension(i)) {
					return false;
				}
			}
			return true;
		}catch(Exception e) {
			return false;
		}
	}
	
	@Override
	public String toString() {
		String str = "(";
		for(int i = 0; i < values.size(); i++) {
			str += values.get(i);
			if(i != values.size()-1) {
				str += ", ";
			}
		}
		return str + ")";
	}
}

File: src/Vector.cpp
/**
 * @file Vector.cpp
 * Represents an N-dimensional vector.
 * @author Conlan Wesson
 */

#include "Vector.h"
#include <cmath>

namespace Funge {

Vector::Vector(){
	
}

Vector::Vector(const Vector& orig) : values(orig.values){
	normalize();
}

Vector::Vector(std::initializer_list<dim_t> v) : values(v){
	normalize();
}

Vector& Vector::operator=(const Vector& orig){
	values.clear();
	values = orig.values;
	normalize();
	return *this;
}

Vector& Vector::operator=(std::initializer_list<dim_t> orig){
	values.clear();
	values = orig;
	normalize();
	return *this;
}

void Vector::normalize(){
	while(values.size() > 1 && values.back() == 0){
		values.pop_back();
	}
}

size_t Vector::size() const{
	return values.size();
}

dim_t Vector::get(size_t d) const{
	dim_t ret = 0;
	if(d < values.size()){
		ret = values[d];
	}
	return ret;
}

void Vector::set(size_t d, dim_t v){
	while(d >= values.size()){
		values.push_back(0);
	}
	values[d] = v;
	normalize();
}

void Vector::reverse(){
	for(size_t i = 0; i < values.size(); ++i){
		set(i, -get(i));
	}
}

void Vector::left(){
	dim_t tx = get(0);
	dim_t ty = get(1);
	set(0, ty);
	set(1, -tx);
}

void Vector::right(){
	dim_t tx = get(0);
	dim_t ty = get(1);
	set(0, -ty);
	set(1, tx);
}

dim_t Vector::operator[](size_t x) const{
	return this->get(x);
}

Vector& Vector::operator+=(const Vector& other){
	size_t len = std::max(this->values.size(), other.values.size());
	for(size_t i = 0; i < len; ++i){
		this->set(i, this->get(i)+other.get(i));
	}
	return *this;
}

Vector operator+(const Vector& lhs, const Vector& rhs) {
	return Vector(lhs) += rhs;
}

Vector& Vector::operator-=(const Vector& other){
	size_t len = std::max(this->values.size(), other.values.size());
	for(size_t i = 0; i < len; ++i){
		this->set(i, this->get(i)-other.get(i));
	}
	return *this;
}

Vector operator-(const Vector& lhs, const Vector& rhs) {
	return Vector(lhs) -= rhs;
}

bool Vector::operator<(const Vector& other) const{
	size_t len = std::max(this->values.size(), other.values.size());
	for(size_t i = len; i > 0; --i){
		dim_t lhs = this->get(i-1);
		dim_t rhs = other.get(i-1);
		if(lhs > rhs){
			return false;
		}else if(lhs < rhs){
			return true;
		}
	}
	return false;
}

bool Vector::operator<=(const Vector& other) const{
	size_t len = std::max(this->values.size(), other.values.size());
	for(size_t i = len; i > 0; --i){
		dim_t lhs = this->get(i-1);
		dim_t rhs = other.get(i-1);
		if(lhs > rhs){
			return false;
		}else if(lhs < rhs){
			return true;
		}
	}
	return true;
}

bool Vector::operator==(const Vector& other) const{
	size_t len = std::max(this->values.size(), other.values.size());
	for(size_t i = 0; i < len; ++i){
		if(this->get(i) != other.get(i)){
			return false;
		}
	}
	return true;
}

bool Vector::operator!=(const Vector& other) const{
	return !(*this == other);
}

std::ostream& operator<<(std::ostream& os, const Vector& rhs){
	os << "(";
	for(size_t i = 0; i < rhs.values.size(); ++i){
		os << rhs.values[i];
		if(i != rhs.values.size()-1){
			os << ", ";
		}
	}
	os << ")";
	return os;
}

std::istream& operator>>(std::istream& is, Vector& rhs){
	rhs.values.clear();
	char c;
	is >> c;
	if(c == '('){
		is >> c;
		size_t d = 0;
		dim_t v = 0;
		bool neg = false;
		while(c != ')'){
			if(isdigit(c)){
				v = (v*10)+(c-'0');
			}else if(c == ','){
				if(neg){
					v = -v;
				}
				rhs.set(d, v);
				++d;
				v = 0;
				neg = false;
			}else if(v == 0 && c == '-'){
				neg = true;
			}else if(!isspace(c)){
				is.unget();
				break;
			}
			is >> c;
		}
		if(neg){
			v = -v;
		}
		rhs.set(d, v);
	}else{
		is.unget();
	}
	rhs.normalize();
	return is;
}

}
